[TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63587
* #63586
* __->__ #63216
* #63778

Currently there are three classes managed by KernelArena: Expr, Stmt,
and Tensor (and derived classes). KernelArena has been a long standing
painpoint for NNC devs and we're moving away from that memory management
model to ref-count based memory model (using shared_ptr). This commit
switches Expr and Stmt to shared_ptr and is the biggest change in this
transition. Later commits will detach Tensor from KernelArena and kill
the arena + scope altogether.

Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

cc @gmagogsfm

Commit:
[WIP][TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr.

[ghstack-poisoned]

Commit:
rebase on "[WIP][TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."




[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
rebase on "[WIP][TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
rebase on "[WIP][TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
Update on "[WIP][TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
Update on "[WIP][TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
Update on "[WIP][TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
Update on "[WIP][TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
Update on "[WIP][TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
Update on "[WIP][TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."




Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
Update on "[TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Currently there are three classes managed by KernelArena: Expr, Stmt,
and Tensor (and derived classes). KernelArena has been a long standing
painpoint for NNC devs and we're moving away from that memory management
model to ref-count based memory model (using shared_ptr). This commit
switches Expr and Stmt to shared_ptr and is the biggest change in this
transition. Later commits will detach Tensor from KernelArena and kill
the arena + scope altogether.

Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
Update on "[TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Currently there are three classes managed by KernelArena: Expr, Stmt,
and Tensor (and derived classes). KernelArena has been a long standing
painpoint for NNC devs and we're moving away from that memory management
model to ref-count based memory model (using shared_ptr). This commit
switches Expr and Stmt to shared_ptr and is the biggest change in this
transition. Later commits will detach Tensor from KernelArena and kill
the arena + scope altogether.

Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
Update on "[TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Currently there are three classes managed by KernelArena: Expr, Stmt,
and Tensor (and derived classes). KernelArena has been a long standing
painpoint for NNC devs and we're moving away from that memory management
model to ref-count based memory model (using shared_ptr). This commit
switches Expr and Stmt to shared_ptr and is the biggest change in this
transition. Later commits will detach Tensor from KernelArena and kill
the arena + scope altogether.

Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
rebase on "[TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Currently there are three classes managed by KernelArena: Expr, Stmt,
and Tensor (and derived classes). KernelArena has been a long standing
painpoint for NNC devs and we're moving away from that memory management
model to ref-count based memory model (using shared_ptr). This commit
switches Expr and Stmt to shared_ptr and is the biggest change in this
transition. Later commits will detach Tensor from KernelArena and kill
the arena + scope altogether.

Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
Update on "[TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Currently there are three classes managed by KernelArena: Expr, Stmt,
and Tensor (and derived classes). KernelArena has been a long standing
painpoint for NNC devs and we're moving away from that memory management
model to ref-count based memory model (using shared_ptr). This commit
switches Expr and Stmt to shared_ptr and is the biggest change in this
transition. Later commits will detach Tensor from KernelArena and kill
the arena + scope altogether.

Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
rebase on "[TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Currently there are three classes managed by KernelArena: Expr, Stmt,
and Tensor (and derived classes). KernelArena has been a long standing
painpoint for NNC devs and we're moving away from that memory management
model to ref-count based memory model (using shared_ptr). This commit
switches Expr and Stmt to shared_ptr and is the biggest change in this
transition. Later commits will detach Tensor from KernelArena and kill
the arena + scope altogether.

Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

Commit:
Update on "[TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Currently there are three classes managed by KernelArena: Expr, Stmt,
and Tensor (and derived classes). KernelArena has been a long standing
painpoint for NNC devs and we're moving away from that memory management
model to ref-count based memory model (using shared_ptr). This commit
switches Expr and Stmt to shared_ptr and is the biggest change in this
transition. Later commits will detach Tensor from KernelArena and kill
the arena + scope altogether.

Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).

navahgar commented on torch/csrc/jit/tensorexpr/fwd_decls.h:
```diff
@@ -1,26 +1,27 @@
 
 template <typename Node>
-using NodePtr = Node*;
+using NodePtr = std::shared_ptr<Node>;
```
Is it possible to keep this change, i.e., switching to `shared_ptr` alone separate from all the other changes in the diff? If there is any problem and you have to revert, you will only revert a much smaller diff that way. 

Just a suggestion...

ZolotukhinM replied:
Good point. Initially I did that, but then CI detected some failures in the code I wasn't testing locally so some more prep code changes creeped in into this PR. I split that into #63778. This PR can't be reduced to a single line change, but overall now it should be reviewable and mostly contain the relevant to shared_ptr switch changes.

navahgar reviewed (approved):
LGTM. 

One suggestion below.

Commit:
Update on "[TensorExpr] Switch Exprs and Stmt from kernel-arena to shared_ptr."


Currently there are three classes managed by KernelArena: Expr, Stmt,
and Tensor (and derived classes). KernelArena has been a long standing
painpoint for NNC devs and we're moving away from that memory management
model to ref-count based memory model (using shared_ptr). This commit
switches Expr and Stmt to shared_ptr and is the biggest change in this
transition. Later commits will detach Tensor from KernelArena and kill
the arena + scope altogether.

Differential Revision: [D30353195](https://our.internmc.facebook.com/intern/diff/D30353195)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30353195).
